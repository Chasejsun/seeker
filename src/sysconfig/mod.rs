mod net;
mod proc;

pub use net::{setup_ip, DNSSetup};
pub use proc::sys::{list_system_proc_socks, list_user_proc_socks};
