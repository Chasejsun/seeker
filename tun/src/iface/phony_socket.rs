use smoltcp::phy::{Device, DeviceCapabilities};
use smoltcp::storage::RingBuffer;
use smoltcp::time::Instant;
use smoltcp::Result;
use smoltcp::{phy, Error};
use tracing::debug;

const MAX_PACKETS: usize = 1024;

#[derive(Debug)]
pub struct PhonySocket {
    mtu: usize,
    rx: RingBuffer<'static, Vec<u8>>,
    tx: RingBuffer<'static, Vec<u8>>,
}

fn enqueue<'a>(
    ring_buffer: &'a mut RingBuffer<'static, Vec<u8>>,
    mtu: usize,
) -> Result<&'a mut Vec<u8>> {
    let buf = ring_buffer.enqueue_one()?;
    buf.resize(mtu, 0);
    Ok(buf)
}

fn enqueue_many<'a>(
    ring_buffer: &'a mut RingBuffer<'static, Vec<u8>>,
    mtu: usize,
) -> &'a mut [Vec<u8>] {
    let bufs = ring_buffer.enqueue_many(10);
    for buf in bufs.iter_mut() {
        buf.resize(mtu, 0);
    }
    bufs
}

fn deque<'a>(ring_buffer: &'a mut RingBuffer<'static, Vec<u8>>) -> Option<&'a mut Vec<u8>> {
    //    NLL 目前没法支持这种类型的代码，只能写出这样来绕过 borrow checker
    //    loop {
    //        let buf = ring_buffer.dequeue_one()?;
    //        if !buf.is_empty() {
    //            return Some(buf);
    //        }
    //    }

    loop {
        let buf = ring_buffer.dequeue_one_with(|buf| {
            if !buf.is_empty() {
                Err(smoltcp::Error::Checksum)
            } else {
                Ok(buf)
            }
        });
        match buf {
            Ok(_) => {}
            Err(Error::Checksum) => break,
            Err(Error::Exhausted) => return None,
            Err(_) => unreachable!(),
        }
    }
    match ring_buffer.dequeue_one() {
        Ok(buf) => Some(buf),
        Err(Error::Exhausted) => None,
        Err(_) => unreachable!(),
    }
}

impl PhonySocket {
    pub fn new(mtu: usize) -> Self {
        let rx: Vec<_> = (0..MAX_PACKETS).map(|_| vec![0; mtu]).collect();
        let tx: Vec<_> = (0..MAX_PACKETS).map(|_| vec![0; mtu]).collect();
        PhonySocket {
            mtu,
            rx: RingBuffer::new(rx),
            tx: RingBuffer::new(tx),
        }
    }

    pub fn populate_rx(&mut self) -> Option<&mut Vec<u8>> {
        enqueue(&mut self.rx, self.mtu).ok()
    }

    pub fn populate_rx_many(&mut self) -> &mut [Vec<u8>] {
        enqueue_many(&mut self.rx, self.mtu)
    }

    pub fn vacate_tx(&mut self) -> Option<&mut Vec<u8>> {
        deque(&mut self.tx)
    }
}

impl<'a> Device<'a> for PhonySocket {
    type RxToken = RxToken<'a>;
    type TxToken = TxToken<'a>;

    fn receive(&'a mut self) -> Option<(Self::RxToken, Self::TxToken)> {
        let buf = deque(&mut self.rx)?;
        let rx = RxToken { buf };
        let tx = TxToken {
            ring_buffer: &mut self.tx,
        };
        Some((rx, tx))
    }

    fn transmit(&'a mut self) -> Option<Self::TxToken> {
        let tx = TxToken {
            ring_buffer: &mut self.tx,
        };
        Some(tx)
    }

    fn capabilities(&self) -> DeviceCapabilities {
        let mut cap = DeviceCapabilities::default();
        cap.max_transmission_unit = self.mtu;
        cap
    }
}

#[doc(hidden)]
pub struct TxToken<'a> {
    ring_buffer: &'a mut RingBuffer<'static, Vec<u8>>,
}

#[doc(hidden)]
pub struct RxToken<'a> {
    buf: &'a mut [u8],
}

impl<'a> phy::TxToken for TxToken<'a> {
    fn consume<R, F>(self, _timestamp: Instant, len: usize, f: F) -> smoltcp::Result<R>
    where
        F: FnOnce(&mut [u8]) -> smoltcp::Result<R>,
    {
        let buf = enqueue(self.ring_buffer, len)?;
        buf.resize(len, 0);
        let ret = f(buf);
        debug!("TxToken.consume {} bytes", len);
        ret
    }
}

impl<'a> phy::RxToken for RxToken<'a> {
    fn consume<R, F>(self, _timestamp: Instant, f: F) -> smoltcp::Result<R>
    where
        F: FnOnce(&mut [u8]) -> smoltcp::Result<R>,
    {
        f(self.buf)
    }
}
