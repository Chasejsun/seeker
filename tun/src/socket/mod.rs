use smoltcp::wire::{IpAddress, IpEndpoint};
use std::net::{Ipv4Addr, SocketAddr};

fn to_socket_addr(endpoint: IpEndpoint) -> SocketAddr {
    match endpoint.addr {
        IpAddress::Ipv4(addr) => {
            let a: Ipv4Addr = addr.into();
            (a, endpoint.port).into()
        }
        _ => unreachable!(),
    }
}

mod tcp_socket;
mod udp_socket;

use smoltcp::socket::SocketHandle;
use std::fmt;
use std::fmt::Display;
pub use tcp_socket::TunTcpSocket;
pub use udp_socket::TunUdpSocket;

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum TunSocket {
    Tcp(TunTcpSocket),
    Udp(TunUdpSocket),
}

impl TunSocket {
    pub unsafe fn new_tcp_socket(handle: SocketHandle) -> TunSocket {
        TunSocket::Tcp(TunTcpSocket::new(handle))
    }

    pub unsafe fn new_udp_socket(handle: SocketHandle) -> TunSocket {
        TunSocket::Udp(TunUdpSocket::new(handle))
    }

    pub fn handle(&self) -> SocketHandle {
        match self {
            TunSocket::Tcp(s) => s.handle(),
            TunSocket::Udp(s) => s.handle(),
        }
    }

    pub fn local_addr(&self) -> SocketAddr {
        match self {
            TunSocket::Tcp(s) => s.local_addr(),
            TunSocket::Udp(s) => s.local_addr(),
        }
    }
}

impl Display for TunSocket {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}
