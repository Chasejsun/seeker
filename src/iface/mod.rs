use crate::phy::TunSocket;
use mio::unix::EventedFd;
use mio::{Evented, Poll, PollOpt, Ready, Token};
use smoltcp::socket::SocketSet;
use std::io;
use std::io::{Read, Write};
use std::os::unix::io::AsRawFd;

pub mod ethernet;

pub struct Iface {
    inner: ethernet::Interface<'static, TunSocket>,
    sockets: SocketSet<'static, 'static, 'static>,
}

impl Iface {
    pub fn new(inner: ethernet::Interface<'static, TunSocket>) -> Self {
        let sockets = SocketSet::new(vec![]);
        Iface { inner, sockets }
    }
}

impl Evented for Iface {
    fn register(
        &self,
        poll: &Poll,
        token: Token,
        interest: Ready,
        opts: PollOpt,
    ) -> io::Result<()> {
        EventedFd(&self.inner.device().as_raw_fd()).register(poll, token, interest, opts)
    }

    fn reregister(
        &self,
        poll: &Poll,
        token: Token,
        interest: Ready,
        opts: PollOpt,
    ) -> io::Result<()> {
        EventedFd(&self.inner.device().as_raw_fd()).reregister(poll, token, interest, opts)
    }

    fn deregister(&self, poll: &Poll) -> io::Result<()> {
        EventedFd(&self.inner.device().as_raw_fd()).deregister(poll)
    }
}

impl Read for Iface {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        unimplemented!()
    }
}

impl Write for Iface {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        unimplemented!()
    }

    fn flush(&mut self) -> io::Result<()> {
        unimplemented!()
    }
}
