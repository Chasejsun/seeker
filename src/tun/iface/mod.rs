use super::phy::TunSocket;
use mio::unix::EventedFd;
use mio::{Evented, Poll, PollOpt, Ready, Token};
use phony_socket::PhonySocket;
use smoltcp::socket::SocketSet;
use std::io;
use std::io::{Read, Write};
use std::os::unix::io::AsRawFd;

pub mod ethernet;
pub mod phony_socket;
