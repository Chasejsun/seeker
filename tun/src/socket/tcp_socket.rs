use crate::socket::to_socket_addr;
use crate::TUN;
use futures::{AsyncRead, AsyncWrite};
use smoltcp::socket::{SocketHandle, TcpSocket};
use std::io;
use std::io::Error;
use std::net::SocketAddr;
use std::pin::Pin;
use std::task::{Context, Poll};
use std::time::Duration;
use tracing::{debug, info};

#[derive(Debug, Eq, PartialEq, Hash)]
pub struct TunTcpSocket {
    handle: SocketHandle,
}

impl TunTcpSocket {
    pub unsafe fn new(handle: SocketHandle) -> Self {
        debug!("TunTcpSocket.new: {}", handle);

        TunTcpSocket { handle }
    }

    pub fn local_addr(&self) -> SocketAddr {
        let mut guard = TUN.try_lock_for(Duration::from_secs(1)).unwrap();
        let mut_tun = guard.as_mut().expect("no tun setup");
        let socket = mut_tun.sockets.get::<TcpSocket>(self.handle);
        to_socket_addr(socket.local_endpoint())
    }

    pub fn remote_addr(&self) -> SocketAddr {
        let mut guard = TUN.try_lock_for(Duration::from_secs(1)).unwrap();
        let mut_tun = guard.as_mut().expect("no tun setup");
        let socket = mut_tun.sockets.get::<TcpSocket>(self.handle);
        to_socket_addr(socket.remote_endpoint())
    }

    pub fn handle(&self) -> SocketHandle {
        self.handle
    }
}

impl Clone for TunTcpSocket {
    fn clone(&self) -> Self {
        debug!("TunTcpSocket.clone: {}", self.handle);

        let mut guard = TUN.try_lock_for(Duration::from_secs(1)).unwrap();
        let mut_tun = guard.as_mut().expect("no tun setup");
        mut_tun.sockets.retain(self.handle);

        TunTcpSocket {
            handle: self.handle,
        }
    }
}

impl Drop for TunTcpSocket {
    fn drop(&mut self) {
        debug!("TunTcpSocket.drop: {}", self.handle);
        let mut guard = TUN.try_lock_for(Duration::from_secs(1)).unwrap();
        let mut_tun = guard.as_mut().expect("no tun setup");

        mut_tun.sockets.release(self.handle);
    }
}

impl AsyncRead for TunTcpSocket {
    fn poll_read(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &mut [u8],
    ) -> Poll<Result<usize, Error>> {
        let mut guard = TUN.try_lock_for(Duration::from_secs(1)).unwrap();
        let mut_tun = guard.as_mut().expect("no tun setup");
        let mut socket = mut_tun.sockets.get::<TcpSocket>(self.handle);
        debug!("TunTcpSocket.read socket state: {}", socket.state());
        if socket.may_recv() {
            if socket.can_recv() {
                let size = socket
                    .recv_slice(buf)
                    .map_err(|_| -> io::Error { io::ErrorKind::Other.into() })?;
                debug!(
                    "TunTcpSocket.read recv {} bytes",
                    size,
                    //                        std::str::from_utf8(buf).unwrap()
                );
                assert!(size > 0);
                Poll::Ready(Ok(size))
            } else {
                debug!("TunTcpSocket.read will block");
                let h = socket.handle();
                mut_tun
                    .socket_read_tasks
                    .insert(h, Some(cx.waker().clone()));
                Poll::Pending
            }
        } else {
            info!("read eof for tcp socket: {}", self.handle);
            Poll::Ready(Ok(0))
        }
    }
}

impl AsyncWrite for TunTcpSocket {
    fn poll_write(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
        buf: &[u8],
    ) -> Poll<Result<usize, Error>> {
        debug!("TunTcpSocket.write");
        let mut guard = TUN.try_lock_for(Duration::from_secs(1)).unwrap();
        let mut_tun = guard.as_mut().expect("no tun setup");
        let mut socket = mut_tun.sockets.get::<TcpSocket>(self.handle);
        if socket.may_send() {
            if socket.can_send() {
                let size = socket
                    .send_slice(buf)
                    .map_err(|_| -> io::Error { io::ErrorKind::Other.into() })?;
                debug!(
                    "TunTcpSocket.write send {} bytes",
                    size,
                    //                        std::str::from_utf8(buf).unwrap()
                );

                if let Some(waker) = mut_tun.tun_write_task.take() {
                    waker.wake();
                }
                Poll::Ready(Ok(size))
            } else {
                debug!("TunTcpSocket.write will block");
                let h = socket.handle();
                mut_tun
                    .socket_write_tasks
                    .insert(h, Some(cx.waker().clone()));
                Poll::Pending
            }
        } else {
            info!("write eof for tcp socket: {}", self.handle);
            Poll::Ready(Ok(0))
        }
    }

    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
        debug!("flush");
        Poll::Ready(Ok(()))
    }

    fn poll_close(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
        debug!("close");
        Poll::Ready(Ok(()))
    }
}
