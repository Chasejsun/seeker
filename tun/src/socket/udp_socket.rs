use crate::socket::to_socket_addr;
use crate::TUN;
use async_std::future::poll_fn;
use smoltcp::socket::{SocketHandle, UdpSocket};
use std::io;
use std::io::Result;
use std::net::SocketAddr;
use std::task::{Context, Poll};
use std::time::Duration;
use tracing::debug;

#[derive(Debug, Eq, PartialEq, Hash)]
pub struct TunUdpSocket {
    handle: SocketHandle,
}

impl TunUdpSocket {
    /// # Safety
    ///
    /// You need to make sure handle have at lease one reference.
    pub unsafe fn new(handle: SocketHandle) -> Self {
        debug!("TunUdpSocket.new: {}", handle);
        TunUdpSocket { handle }
    }

    pub fn handle(&self) -> SocketHandle {
        self.handle
    }

    pub fn local_addr(&self) -> SocketAddr {
        let mut guard = TUN.lock();
        let mut_tun = guard.as_mut().expect("no tun setup");
        let socket = mut_tun.sockets.get::<UdpSocket>(self.handle);
        to_socket_addr(socket.endpoint())
    }

    pub fn poll_recv_from(
        &self,
        cx: &mut Context<'_>,
        buf: &mut [u8],
    ) -> Poll<Result<(usize, SocketAddr)>> {
        debug!("TunUdpSocket.read");
        let mut guard = TUN.lock();
        let mut_tun = guard.as_mut().expect("no tun setup");
        let mut socket = mut_tun.sockets.get::<UdpSocket>(self.handle);
        if socket.can_recv() {
            let (size, endpoint) = socket
                .recv_slice(buf)
                .map_err(|_| -> io::Error { io::ErrorKind::Other.into() })?;
            debug!("TunUdpSocket.read {} bytes", size);
            Poll::Ready(Ok((size, to_socket_addr(endpoint))))
        } else {
            let h = socket.handle();
            mut_tun
                .socket_read_tasks
                .insert(h, Some(cx.waker().clone()));
            debug!("TunUdpSocket.read blocks: {:?}", h);
            Poll::Pending
        }
    }

    pub fn poll_send_to(
        &self,
        cx: &mut Context<'_>,
        buf: &[u8],
        target: &SocketAddr,
    ) -> Poll<Result<usize>> {
        debug!("TunUdpSocket.write");
        let mut guard = TUN.lock();
        let mut_tun = guard.as_mut().expect("no tun setup");
        let mut socket = mut_tun.sockets.get::<UdpSocket>(self.handle);
        if socket.can_send() {
            let endpoint = match target {
                SocketAddr::V4(addr) => addr.clone().into(),
                SocketAddr::V6(_) => unreachable!(),
            };
            socket
                .send_slice(buf, endpoint)
                .map_err(|_e| -> io::Error { io::ErrorKind::Other.into() })?;
            if let Some(waker) = mut_tun.tun_write_task.take() {
                waker.wake();
            }
            Poll::Ready(Ok(buf.len()))
        } else {
            let h = socket.handle();
            mut_tun
                .socket_write_tasks
                .insert(h, Some(cx.waker().clone()));
            Poll::Pending
        }
    }

    pub async fn send_to(&self, buf: &[u8], addr: &SocketAddr) -> io::Result<usize> {
        poll_fn(|cx| self.poll_send_to(cx, buf, addr)).await
    }

    pub async fn recv_from(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)> {
        poll_fn(|cx| self.poll_recv_from(cx, buf)).await
    }
}

impl Clone for TunUdpSocket {
    fn clone(&self) -> Self {
        let mut guard = TUN.lock();
        let mut_tun = guard.as_mut().expect("no tun setup");
        mut_tun.sockets.retain(self.handle);
        TunUdpSocket {
            handle: self.handle,
        }
    }
}

impl Drop for TunUdpSocket {
    fn drop(&mut self) {
        debug!("TunUdpSocket.drop: {}", self.handle);
        let mut guard = TUN.lock();
        let mut_tun = guard.as_mut().expect("no tun setup");
        mut_tun.sockets.release(self.handle)
    }
}
