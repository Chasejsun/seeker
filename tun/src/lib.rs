use std::collections::HashMap;
use std::future::Future;
use std::io::Result;
use std::net::Ipv4Addr;
use std::pin::Pin;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::task::{Context, Poll, Waker};
use std::time::Duration;

use futures::{ready, AsyncRead, AsyncWrite, Stream};
use parking_lot::Mutex;
use smoltcp::socket::{Socket, SocketHandle, SocketSet};
use smoltcp::time::Instant;
use smoltcp::wire::Ipv4Cidr;
use tracing::{debug, error, trace};

use iface::ethernet::{Interface, InterfaceBuilder};
use iface::phony_socket::PhonySocket;
use lazy_static::lazy_static;
use sysconfig::setup_ip;

use crate::socket::TunSocket;

pub mod iface;
pub mod phy;

#[macro_use]
pub mod socket;

lazy_static! {
    static ref TUN: Mutex<Option<Tun>> = Mutex::new(None);
}

pub struct Tun {
    iface: Interface<'static, PhonySocket>,
    tun: phy::TunSocket,
    sockets: SocketSet<'static, 'static, 'static>,
    new_sockets: Vec<TunSocket>,
    socket_read_tasks: HashMap<SocketHandle, Option<Waker>>,
    socket_write_tasks: HashMap<SocketHandle, Option<Waker>>,
    tun_write_task: Option<Waker>,
    to_terminate: Arc<AtomicBool>,
}

impl Tun {
    pub fn setup(
        tun_name: String,
        tun_ip: Ipv4Addr,
        tun_cidr: Ipv4Cidr,
        to_terminate: Arc<AtomicBool>,
    ) {
        let tun = phy::TunSocket::new(tun_name.as_str());
        let tun_name = tun.name();
        if cfg!(target_os = "macos") {
            setup_ip(
                tun_name,
                tun_ip.to_string().as_str(),
                tun_cidr.to_string().as_str(),
            );
        } else {
            let new_ip = Ipv4Cidr::from_netmask(tun_ip.into(), tun_cidr.netmask()).unwrap();
            setup_ip(
                tun_name,
                new_ip.to_string().as_str(),
                tun_cidr.to_string().as_str(),
            );
        }

        let device = PhonySocket::new(tun.mtu());
        let ip_addrs = vec![tun_cidr.into()];
        let iface = InterfaceBuilder::new(device)
            .ip_addrs(ip_addrs)
            .any_ip(true)
            .finalize();

        let sockets = SocketSet::new(vec![]);
        let new_sockets = Vec::new();
        let tun = Tun {
            iface,
            tun,
            sockets,
            new_sockets,
            socket_read_tasks: HashMap::new(),
            socket_write_tasks: HashMap::new(),
            tun_write_task: None,
            to_terminate,
        };
        let _ = TUN
            .try_lock_for(Duration::from_secs(1))
            .unwrap()
            .replace(tun);
    }

    pub fn listen() -> TunListen {
        TunListen
    }

    pub fn bg_send() -> TunWrite {
        TunWrite
    }
}

pub struct TunListen;

#[derive(Debug, PartialEq)]
enum Handle {
    Tcp(SocketHandle),
    Udp(SocketHandle),
}

impl TunListen {
    fn may_recv_tun_handles(mut_tun: &mut Tun, handles: &mut Vec<Handle>) {
        for s in mut_tun.sockets.iter() {
            match s {
                Socket::Tcp(s) if s.may_recv() => {
                    handles.push(Handle::Tcp(s.handle()));
                }
                Socket::Udp(s) if s.is_open() => {
                    handles.push(Handle::Udp(s.handle()));
                }
                _ => {}
            }
        }
    }
}

impl Stream for TunListen {
    type Item = Result<TunSocket>;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let mut guard = TUN.try_lock_for(Duration::from_secs(1)).unwrap();
        let mut_tun = guard.as_mut().expect("no tun setup");
        let size = mut_tun.sockets.iter().count();
        let mut before_handle = Vec::with_capacity(size);
        let mut after_handle = Vec::with_capacity(size);

        loop {
            if mut_tun.to_terminate.load(Ordering::Relaxed) {
                return Poll::Ready(None);
            }

            if let Some(s) = mut_tun.new_sockets.pop() {
                trace!("new socket accepted: {}", s);
                return Poll::Ready(Some(Ok(s)));
            }

            TunListen::may_recv_tun_handles(mut_tun, &mut before_handle);

            {
                let phony_socket = mut_tun.iface.device_mut();
                let mut total_size = 0;
                while let Some(buf) = phony_socket.populate_rx() {
                    let size = match Pin::new(&mut mut_tun.tun).poll_read(cx, buf) {
                        Poll::Ready(Ok(size)) => {
                            buf.truncate(size);
                            trace!("tun.poll_read size {}", size);
                            size
                        }
                        Poll::Ready(Err(e)) => return Poll::Ready(Some(Err(e))),
                        Poll::Pending => {
                            buf.clear();
                            if total_size > 0 {
                                trace!("tun.poll_read will block, total read size: {}", total_size);
                                break;
                            } else {
                                return Poll::Pending;
                            }
                        }
                    };
                    total_size += size;
                }
            }

            match mut_tun
                .iface
                .poll_read(&mut mut_tun.sockets, Instant::now())
            {
                Ok(_) => {}
                Err(smoltcp::Error::Malformed) | Err(smoltcp::Error::Dropped) => {}
                Err(e) => {
                    error!("iface.poll_read error: {}, poll again", e);
                }
            };

            if let Some(waker) = mut_tun.tun_write_task.take() {
                debug!("notify tun for write");
                waker.wake();
            }

            for mut socket in mut_tun.sockets.iter_mut() {
                match &mut *socket {
                    Socket::Tcp(ref mut s) => {
                        if s.is_open() && (s.can_recv() || !s.may_recv()) {
                            if let Some(t) = mut_tun.socket_read_tasks.get_mut(&s.handle()) {
                                if let Some(waker) = t.take() {
                                    debug!("notify tcp socket {} for read", s.handle());
                                    waker.wake();
                                }
                            }
                        }
                    }

                    Socket::Udp(s) => {
                        if s.is_open() && s.can_recv() {
                            if let Some(t) = mut_tun.socket_read_tasks.get_mut(&s.handle()) {
                                if let Some(waker) = t.take() {
                                    debug!("notify udp socket {} for read", s.handle());
                                    waker.wake();
                                }
                            }
                        }
                    }
                    _ => unreachable!(),
                }
            }

            mut_tun.sockets.prune();

            TunListen::may_recv_tun_handles(mut_tun, &mut after_handle);
            for handle in &after_handle {
                if !before_handle.contains(handle) {
                    let sock = match handle {
                        Handle::Tcp(h) => unsafe { TunSocket::new_tcp_socket(*h) },
                        Handle::Udp(h) => unsafe { TunSocket::new_udp_socket(*h) },
                    };
                    mut_tun.new_sockets.push(sock);
                }
            }
            before_handle.clear();
            after_handle.clear();
        }
    }
}

pub struct TunWrite;

impl TunWrite {
    fn poll_write_sockets_to_phoney_socket(cx: &mut Context<'_>, tun: &mut Tun) -> Result<bool> {
        let processed_any = match tun.iface.poll_write(&mut tun.sockets, Instant::now()) {
            Ok(any) => any,
            Err(smoltcp::Error::Malformed) | Err(smoltcp::Error::Dropped) => true,
            Err(e) => {
                error!("poll_write error: {}, poll again", e);
                true
            }
        };

        for socket in tun.sockets.iter_mut() {
            match &*socket {
                Socket::Tcp(s) => {
                    //                    debug!("tcp socket {} state: {}.", s.handle(), s.state());
                    if s.can_send() {
                        //                        debug!("tcp socket {} can send", s.handle());
                        if let Some(t) = tun.socket_write_tasks.get_mut(&s.handle()) {
                            if let Some(waker) = t.take() {
                                debug!("notify tcp socket {} for write", s.handle());
                                waker.wake();
                            }
                        }
                    }
                }
                Socket::Udp(s) => {
                    if s.can_send() {
                        //                        debug!("udp socket {} can send.", s.handle());
                        if let Some(t) = tun.socket_read_tasks.get_mut(&s.handle()) {
                            if let Some(waker) = t.take() {
                                debug!("notify udp socket {} for write", s.handle());
                                waker.wake();
                            }
                        }
                    }
                }
                _ => unreachable!(),
            }
        }

        if !processed_any {
            debug!("TunWrite NotReady");
            tun.tun_write_task = Some(cx.waker().clone());
        }
        Ok(processed_any)
    }
}

impl Future for TunWrite {
    type Output = Result<()>;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        loop {
            let mut guard = TUN.try_lock_for(Duration::from_secs(5)).unwrap();
            let mut_tun = guard.as_mut().expect("no tun setup");

            if mut_tun.to_terminate.load(Ordering::Relaxed) {
                return Poll::Ready(Ok(()));
            }

            loop {
                let phony_socket = mut_tun.iface.device_mut();
                let buf = phony_socket.vacate_tx();
                match buf {
                    Some(buf) if buf.is_empty() => {}
                    Some(buf) => {
                        let size = ready!(Pin::new(&mut mut_tun.tun).poll_write(cx, &buf)).unwrap();
                        assert_eq!(size, buf.len());
                        debug!("write {} bytes to tun.", size);
                    }
                    None => break,
                }
            }

            if !TunWrite::poll_write_sockets_to_phoney_socket(cx, mut_tun)? {
                return Poll::Pending;
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use std::net::SocketAddr;

    use async_std::io;
    use async_std::net::TcpStream;
    use async_std::task;
    use futures::{AsyncReadExt, AsyncWriteExt, StreamExt};
    use smoltcp::wire::Ipv4Address;

    use super::*;

    #[test]
    fn test_accept_tcp() {
        let to_terminate = Arc::new(AtomicBool::new(false));
        Tun::setup(
            "utun4".to_string(),
            Ipv4Addr::new(10, 0, 0, 1),
            Ipv4Cidr::new(Ipv4Address::new(10, 0, 0, 0), 24),
            to_terminate.clone(),
        );

        task::block_on(async move {
            task::spawn(Tun::bg_send());

            task::spawn(async move {
                let mut stream = Tun::listen();
                loop {
                    match stream.next().await {
                        Some(Ok(TunSocket::Tcp(mut s))) => {
                            assert_eq!(
                                s.local_addr(),
                                "10.0.0.2:80".parse::<SocketAddr>().unwrap()
                            );
                            let mut buf = vec![0; 1024];
                            let size = s.read(&mut buf).await.unwrap();
                            assert_eq!(size, 5);
                            assert_eq!(&buf[..size], "hello".as_bytes());
                        }
                        _ => panic!(),
                    }
                }
            });

            task::sleep(Duration::from_secs(1)).await;
            let mut stream = io::timeout(Duration::from_secs(1), TcpStream::connect("10.0.0.2:80"))
                .await
                .expect("connect 10.0.0.2:80");
            stream.write_all("hello".as_bytes()).await.unwrap();

            task::sleep(Duration::from_secs(1)).await;
        });
    }
}
