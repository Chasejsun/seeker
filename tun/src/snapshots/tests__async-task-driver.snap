---
source: tun/src/lib.rs
expression: "&s.local_addr()"
---
V4(
    10.0.0.2:80,
)
